# Exploración de tablas 
## Proyecto Final

**Equipo:**
* Malváez Flores Axel Daniel
* Peralta Rionda Gabriel Zadquiel

In [1]:
import pandas as pd
import numpy as np

In [2]:
baInvoices = pd.read_excel('./Tablas/baInvoices.xlsx')
baInvoicesItems = pd.read_excel('./Tablas/baInvoicesItems.xlsx')
baStaff = pd.read_excel('./Tablas/baStaff.xlsx')
baStock = pd.read_excel('./Tablas/baStock.xlsx')

## Perfilado de los datos

Perfil de datos (diagnóstico de los datos en términos de completes, correctes, relevancia, falta de programación de reglas de negocio para asegurar integridad y consistencia en los datos, detección de patrones que no se encuentran en correos electrónicos, direcciones, nombres completos, etc.)

### Funciones de ayuda para perfilar

* Completitud

In [3]:
def completeness_calculator(data):
    """Calculates completeness of my data"""
    completeness = (1 - data.isna().sum() / len(data))
    return completeness

def completeness_report(data):
    """Calculates completeness of a data and returns a report"""
    completeness = completeness_calculator(data)
    completeness_report = pd.DataFrame(completeness, columns=['Completeness'])
    completeness_report['Missing'] = (1 - completeness)
    return completeness_report

### baInvoices

#### Completitud

In [4]:
# Reporte completo
baInvoices_completeness = completeness_report(baInvoices)
baInvoices_completeness.shape

(132, 2)

In [5]:
# Columnas con completitud de 1
baInvoices_completeness_1 = baInvoices_completeness[baInvoices_completeness['Completeness'] == 1]
baInvoices_completeness_1.shape

(79, 2)

In [6]:
# Columnas con completitud de 0
baInvoices_completeness_0 = baInvoices_completeness[baInvoices_completeness['Completeness'] == 0]
baInvoices_completeness_0.shape

(47, 2)

In [7]:
# Columnas con missing diferente de 0 y de 1
baInvoices_miss = baInvoices_completeness[(baInvoices_completeness['Missing'] != 0) & (baInvoices_completeness['Missing'] != 1)]
baInvoices_miss.shape

(6, 2)

Tenemos que de todas nuestras 132 columnas en nuestra tabla "baInvoices", 79 de ellas están completas y son las siguientes:

In [8]:
baInvoices_completeness_1.index.values

array(['InvoiceID', 'InvoiceNumber', 'ClientID', 'Date Creation',
       'Date Invoice', 'Currency', 'Sub Total', 'VAT', 'Total',
       'Total letter', 'Discount', 'DiscountAccepted', 'Exportation',
       'Canceled', 'Language', 'DateDocDelivery', 'PaiementDueDate',
       'StaffID', 'SentToAccounting', 'Deleted', 'Rate', 'BranchID',
       'Branch_InvoiceID', 'Branch_OrderID', 'Branch_ClientID',
       'Master_InvoiceID', 'MustBeSynchronized', 'Ticket',
       'MustBeDeliveredToClient', 'Printed', 'CanceledInvoiceID',
       'NewInvoiceID', 'BatchIDExport', 'PeriodicYesNo',
       'PeriodicDayMonth', 'PeriodicDayWeek', 'PeriodicDayBiMonth1',
       'PeriodicDayBiMonth2', 'PeriodicBasedOnInvoiceID', 'QuotationID',
       'CancelationReason', 'Paid', 'Balance', 'StatusPaid',
       'VersionSync', 'InvoiceIDGlobalized', 'Globalized', 'Consignation',
       'PeriodicPeriod', 'PeriodicListIndex', 'CreatedByStaffID',
       'StatusCFD', 'ApprovalNumber', 'SubTotalVAT1', 'VATPerc1',
      

Finalmente las columnas que no están completas al 100% son:

In [9]:
baInvoices_completeness_0.index.values

array(['PaiementType', 'OrderID', 'DiscountKey', 'SaleOtherBranchID',
       'PeriodicTitle', 'GuideNumber', 'DescriptionJustifyOrigin',
       'Comments', 'PeriodicDateTo', 'FolioCFD', 'ApprovalDate',
       'SubTotalVAT2', 'SubTotalVAT3', 'SubTotalVAT4', 'SubTotalVAT5',
       'VATPerc2', 'VATPerc3', 'VATPerc4', 'VATPerc5', 'PrinterSerial',
       'Title', 'CFDStatusError', 'CFDStatusCancelledError',
       'OrdenCompra', 'LugarEntrega', 'Codigo', 'Codigo2',
       'NumeroRemision', 'Cita', 'TipoBulto', 'CantidadBultos',
       'WalmartFechaOrdenCompra', 'FemsaDivision', 'FemsaFechaDesde',
       'FemsaFechaHasta', 'SubAddenda1', 'PaymentRefCustomer',
       'CFDFormOfPaymentIDs', 'CFDFormOfPayment', 'CFDBankAccounts',
       'AddendaContrareciboFecha', 'AddendaContrareciboNumero',
       'AddendaNumeroDeDepartamento', 'AddendaDiasDePago', 'ValidatedOn',
       'CFDMetodoDePago', 'SeriesInvoiceNumber'], dtype=object)

In [10]:
# Columnas que no tienen completitud de 1 pero que tampoco tienen todos los valores faltantes
baInvoices_miss.index.values

array(['ProjectID', 'PerformanceType', 'DateCancel', 'UserID',
       'FechaRemision', 'FechaEntrega'], dtype=object)

In [11]:
baInvoices_completeness.loc[baInvoices_miss.index.values]

Completeness   Missing
ProjectID            0.999131  0.000869
PerformanceType      0.999131  0.000869
DateCancel           0.000869  0.999131
UserID               0.996905  0.003095
FechaRemision        0.000869  0.999131
FechaEntrega         0.000869  0.999131

Exploracion sobre los que no tienen completitud de 1 pero que tampoco tienen todos los valores faltantes

* ProjectID

In [12]:
baInvoices.loc[:, 'ProjectID'].unique()

array([ 0., nan])

* FechaEntrega y FechaRemision

In [13]:
# Columnas que no tienen completitud de 1 pero que tampoco tienen todos los valores faltantes
df = baInvoices.loc[:,baInvoices_miss.index.values]
df.head()

ProjectID                                    PerformanceType DateCancel   
0        0.0  LA CONTRAPRESTACION SE HACE EN UNA SOLA EXHIBI...        NaN  \
1        0.0  LA CONTRAPRESTACION SE HACE EN UNA SOLA EXHIBI...        NaN   
2        0.0  LA CONTRAPRESTACION SE HACE EN UNA SOLA EXHIBI...        NaN   
3        0.0  LA CONTRAPRESTACION SE HACE EN UNA SOLA EXHIBI...        NaN   
4        0.0  LA CONTRAPRESTACION SE HACE EN UNA SOLA EXHIBI...        NaN   

   UserID FechaRemision FechaEntrega  
0     0.0           NaT          NaT  
1     0.0           NaT          NaT  
2     0.0           NaT          NaT  
3     0.0           NaT          NaT  
4     NaN           NaT          NaT

In [14]:
df[df['FechaEntrega'].isna() == False]

ProjectID PerformanceType DateCancel  UserID FechaRemision FechaEntrega
73           NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
74           NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
75           NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
76           NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
93           NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
4434         NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14100        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14101        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14102        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14421        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14427        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14437        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14447        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14448        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
14464        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
15712        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
15713        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
17052        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
17053        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
25236        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
25247        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
25248        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
25249        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
26962        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30
26963        NaN             NaN   00:00:00     NaN    2002-09-30   2002-09-30

In [15]:
np.all(df[df['FechaRemision'].isna() == False].index.values == df[df['FechaEntrega'].isna() == False].index.values)

True

In [16]:
np.all(df[df['ProjectID'].isna() == True].index.values == df[df['FechaRemision'].isna() == False].index.values)

True

In [17]:
np.all(df[df['ProjectID'].isna() == True].index.values == df[df['DateCancel'].isna() == False].index.values)

True

In [18]:
np.all(df[df['ProjectID'].isna() == True].index.values == df[df['PerformanceType'].isna() == True].index.values)

True

In [19]:
np.all(df[df['ProjectID'].isna() == True].index.values == df[df['UserID'].isna() == True].index.values)

/var/folders/8l/f787_tgj7sg6grwn27dvfx0h0000gn/T/ipykernel_80312/1100673596.py:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  np.all(df[df['ProjectID'].isna() == True].index.values == df[df['UserID'].isna() == True].index.values)


False

* UserID

In [20]:
df['UserID'].unique()

array([ 0., nan])

In [21]:
# Datos que no coinciden con la fecha de entrega y la fecha de creación
baInvoices.loc[df[df['FechaEntrega'].isna() == False].index.values, ['InvoiceID','FechaEntrega', 'Date Creation', 'Date Invoice']]

InvoiceID FechaEntrega       Date Creation Date Invoice
73            74   2002-09-30 2022-01-11 13:11:37   2022-01-11
74            75   2002-09-30 2022-01-11 13:19:18   2022-01-11
75            76   2002-09-30 2022-01-11 13:21:15   2022-01-11
76            77   2002-09-30 2022-01-11 13:26:29   2022-01-11
93            94   2002-09-30 2022-01-11 15:05:18   2022-01-10
4434        4435   2002-09-30 2008-06-13 01:02:26   2022-03-05
14100      14101   2002-09-30 2008-06-13 00:07:21   2022-07-01
14101      14102   2002-09-30 2008-06-13 00:07:58   2022-07-01
14102      14103   2002-09-30 2008-06-13 00:08:29   2022-07-01
14421      14422   2002-09-30 2008-06-13 00:17:52   2022-07-06
14427      14428   2002-09-30 2008-06-13 00:43:20   2022-07-06
14437      14438   2002-09-30 2008-06-13 02:05:02   2022-07-06
14447      14448   2002-09-30 2008-06-13 03:52:59   2022-07-06
14448      14449   2002-09-30 2008-06-13 03:55:21   2022-07-06
14464      14465   2002-09-30 2008-06-13 06:28:19   2022-07-06
15712      15713   2002-09-30 2022-07-23 09:11:39   2022-07-22
15713      15714   2002-09-30 2022-07-23 09:14:03   2022-07-22
17052      17053   2002-09-30 2008-06-13 00:03:47   2022-08-09
17053      17054   2002-09-30 2008-06-13 00:16:30   2022-08-09
25236      25237   2002-09-30 2008-06-13 01:21:04   2022-11-17
25247      25248   2002-09-30 2008-11-17 11:45:13   2022-11-17
25248      25249   2002-09-30 2008-11-17 11:47:56   2022-11-17
25249      25250   2002-09-30 2008-11-17 11:51:34   2022-11-17
26962      26963   2002-09-30 2022-12-07 10:05:28   2022-12-06
26963      26964   2002-09-30 2022-12-07 10:09:49   2022-12-06

In [22]:
# Ordenamos por fecha de creación
baInvoices.sort_values(by='Date Creation', ascending=True).head()

InvoiceID  InvoiceNumber  ClientID       Date Creation Date Invoice   
17052      17053          17050         1 2008-06-13 00:03:47   2022-08-09  \
14100      14101          14099         1 2008-06-13 00:07:21   2022-07-01   
14101      14102          14100         1 2008-06-13 00:07:58   2022-07-01   
14102      14103          14101         1 2008-06-13 00:08:29   2022-07-01   
17053      17054          17051         1 2008-06-13 00:16:30   2022-08-09   

      Currency  Sub Total  VAT  Total   
17052      MXN       79.0    0   79.0  \
14100      MXN       79.0    0   79.0   
14101      MXN       48.0    0   48.0   
14102      MXN      384.0    0  384.0   
17053      MXN       44.0    0   44.0   

                                           Total letter  ...   
17052               (SETENTA Y NUEVE PESOS 00/100 M.N.)  ...  \
14100               (SETENTA Y NUEVE PESOS 00/100 M.N.)  ...   
14101               (CUARENTA Y OCHO PESOS 00/100 M.N.)  ...   
14102  (TRESCIENTOS OCHENTA Y CUATRO PESOS 00/100 M.N.)  ...   
17053             (CUARENTA Y CUATRO PESOS 00/100 M.N.)  ...   

       CFDFormOfPayment  CFDBankAccounts  AddendaContrareciboFecha   
17052               NaN              NaN                       NaN  \
14100               NaN              NaN                       NaN   
14101               NaN              NaN                       NaN   
14102               NaN              NaN                       NaN   
17053               NaN              NaN                       NaN   

       AddendaContrareciboNumero  AddendaNumeroDeDepartamento   
17052                        NaN                          NaN  \
14100                        NaN                          NaN   
14101                        NaN                          NaN   
14102                        NaN                          NaN   
17053                        NaN                          NaN   

       AddendaDiasDePago  ValidatedOn  ValidatedBy CFDMetodoDePago   
17052                NaN          NaN            0             NaN  \
14100                NaN          NaN            0             NaN   
14101                NaN          NaN            0             NaN   
14102                NaN          NaN            0             NaN   
17053                NaN          NaN            0             NaN   

      SeriesInvoiceNumber  
17052                 NaN  
14100                 NaN  
14101                 NaN  
14102                 NaN  
17053                 NaN  

[5 rows x 132 columns]

* Exploración de las columnas con completitud de 1

In [23]:
# Datos en los que la fecha de creación es diferente a la fecha de la factura
df = baInvoices[baInvoices['Date Creation'].dt.date != baInvoices['Date Invoice'].dt.date]
print(len(df))
df.head()

21


InvoiceID  InvoiceNumber  ClientID       Date Creation Date Invoice   
93            94             94         1 2022-01-11 15:05:18   2022-01-10  \
4434        4435           4435         1 2008-06-13 01:02:26   2022-03-05   
14100      14101          14099         1 2008-06-13 00:07:21   2022-07-01   
14101      14102          14100         1 2008-06-13 00:07:58   2022-07-01   
14102      14103          14101         1 2008-06-13 00:08:29   2022-07-01   

      Currency  Sub Total  VAT  Total   
93         MXN      184.0    0  184.0  \
4434       MXN       25.0    0   25.0   
14100      MXN       79.0    0   79.0   
14101      MXN       48.0    0   48.0   
14102      MXN      384.0    0  384.0   

                                           Total letter  ...   
93          (CIENTO OCHENTA Y CUATRO PESOS 00/100 M.N.)  ...  \
4434                    (VEINTICINCO PESOS 00/100 M.N.)  ...   
14100               (SETENTA Y NUEVE PESOS 00/100 M.N.)  ...   
14101               (CUARENTA Y OCHO PESOS 00/100 M.N.)  ...   
14102  (TRESCIENTOS OCHENTA Y CUATRO PESOS 00/100 M.N.)  ...   

       CFDFormOfPayment  CFDBankAccounts  AddendaContrareciboFecha   
93                  NaN              NaN                       NaN  \
4434                NaN              NaN                       NaN   
14100               NaN              NaN                       NaN   
14101               NaN              NaN                       NaN   
14102               NaN              NaN                       NaN   

       AddendaContrareciboNumero  AddendaNumeroDeDepartamento   
93                           NaN                          NaN  \
4434                         NaN                          NaN   
14100                        NaN                          NaN   
14101                        NaN                          NaN   
14102                        NaN                          NaN   

       AddendaDiasDePago  ValidatedOn  ValidatedBy CFDMetodoDePago   
93                   NaN          NaN            0             NaN  \
4434                 NaN          NaN            0             NaN   
14100                NaN          NaN            0             NaN   
14101                NaN          NaN            0             NaN   
14102                NaN          NaN            0             NaN   

      SeriesInvoiceNumber  
93                    NaN  
4434                  NaN  
14100                 NaN  
14101                 NaN  
14102                 NaN  

[5 rows x 132 columns]

In [24]:
# Dropeamos los datos del dataframe anterior
x = baInvoices.drop(df.index.values, axis=0)
x[x['Date Creation'].dt.hour == 3]

InvoiceID  InvoiceNumber  ClientID       Date Creation Date Invoice   
10286      10287          10285         1 2022-05-11 03:05:24   2022-05-11  \
10287      10288          10286         1 2022-05-11 03:05:42   2022-05-11   
10288      10289          10287         1 2022-05-11 03:08:57   2022-05-11   
10289      10290          10288         1 2022-05-11 03:15:28   2022-05-11   

      Currency  Sub Total  VAT  Total   
10286      MXN      877.0    0  877.0  \
10287      MXN      130.0    0  130.0   
10288      MXN       83.0    0   83.0   
10289      MXN      151.0    0  151.0   

                                          Total letter  ...  CFDFormOfPayment   
10286  (OCHOCIENTOS SETENTA Y SIETE PESOS 00/100 M.N.)  ...               NaN  \
10287               (CIENTO TREINTA PESOS 00/100 M.N.)  ...               NaN   
10288               (OCHENTA Y TRES PESOS 00/100 M.N.)  ...               NaN   
10289        (CIENTO CINCUENTA Y UN PESOS 00/100 M.N.)  ...               NaN   

       CFDBankAccounts  AddendaContrareciboFecha  AddendaContrareciboNumero   
10286              NaN                       NaN                        NaN  \
10287              NaN                       NaN                        NaN   
10288              NaN                       NaN                        NaN   
10289              NaN                       NaN                        NaN   

       AddendaNumeroDeDepartamento  AddendaDiasDePago  ValidatedOn   
10286                          NaN                NaN          NaN  \
10287                          NaN                NaN          NaN   
10288                          NaN                NaN          NaN   
10289                          NaN                NaN          NaN   

       ValidatedBy CFDMetodoDePago SeriesInvoiceNumber  
10286            0             NaN                 NaN  
10287            0             NaN                 NaN  
10288            0             NaN                 NaN  
10289            0             NaN                 NaN  

[4 rows x 132 columns]

* Análisis de los precios

In [25]:
np.all(baInvoices['VAT'] == 0)

True

Dado que el impuesto en todos los registros es 0, debemos verificar que el subtotal sea igual al total

In [26]:
baInvoices['Sub Total'].equals(baInvoices['Total'])

True

#### Relevancia

Como habíamos visto que las columnas que no tenían completitud exacta de 1 o de 0, no nos aportan gran valor procederemos a omitirlas

In [27]:
baInvoices_cleaned = baInvoices.drop(baInvoices_miss.index.values, axis=1)
baInvoices_cleaned = baInvoices_cleaned.drop(baInvoices_completeness_0.index.values, axis=1)

In [28]:
baInvoices_cleaned.shape
baInvoices_cleaned.head()

InvoiceID  InvoiceNumber  ClientID       Date Creation Date Invoice   
0          1              1         1 2022-01-11 10:15:06   2022-01-11  \
1          2              2         1 2022-01-11 10:29:02   2022-01-11   
2          3              3         1 2022-01-10 10:43:53   2022-01-10   
3          4              4         1 2022-01-10 10:44:34   2022-01-10   
4          5              5         1 2022-01-10 10:44:57   2022-01-10   

  Currency  Sub Total  VAT  Total                           Total letter  ...   
0      MXN       29.0    0   29.0        (VEINTINUEVE PESOS 00/100 M.N.)  ...  \
1      MXN       43.0    0   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)  ...   
2      MXN       61.0    0   61.0       (SESENTA Y UN PESOS 00/100 M.N.)  ...   
3      MXN       57.0    0   57.0  (CINCUENTA Y SIETE PESOS 00/100 M.N.)  ...   
4      MXN       43.0    0   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)  ...   

   DocumentCFDID  WSServerDocumentID  LocalTaxPerc  LocalTaxAmount   
0              0                   0             0               0  \
1              0                   0             0               0   
2              0                   0             0               0   
3              0                   0             0               0   
4              0                   0             0               0   

   LocalTaxType Donation FormOfPaymentPartialities   
0             0    False                     False  \
1             0    False                     False   
2             0    False                     False   
3             0    False                     False   
4             0    False                     False   

   FormOfPaymentPartialityNbr  FormOfPaymentPartialityTotal  ValidatedBy  
0                           0                             0            0  
1                           0                             0            0  
2                           0                             0            0  
3                           0                             0            0  
4                           0                             0            0  

[5 rows x 79 columns]

In [29]:
# Function that allows us to drop columns with equal values or that
# have a high percentage of having the same value in all the records
def drop_equal_values(data, p):
    """Drops columns with equal values"""
    data = data.drop(data.columns[(data.apply(lambda x: x.value_counts().iloc[0]/len(data)) >= p)], axis=1)
    return data

In [30]:
baInvoices_cleaned = drop_equal_values(baInvoices_cleaned, 0.70)
baInvoices_cleaned.shape

(28754, 13)

In [31]:
baInvoices_cleaned.head()

InvoiceID  InvoiceNumber       Date Creation Date Invoice  Sub Total   
0          1              1 2022-01-11 10:15:06   2022-01-11       29.0  \
1          2              2 2022-01-11 10:29:02   2022-01-11       43.0   
2          3              3 2022-01-10 10:43:53   2022-01-10       61.0   
3          4              4 2022-01-10 10:44:34   2022-01-10       57.0   
4          5              5 2022-01-10 10:44:57   2022-01-10       43.0   

   Total                           Total letter DateDocDelivery   
0   29.0        (VEINTINUEVE PESOS 00/100 M.N.)      2022-01-11  \
1   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)      2022-01-11   
2   61.0       (SESENTA Y UN PESOS 00/100 M.N.)      2022-01-10   
3   57.0  (CINCUENTA Y SIETE PESOS 00/100 M.N.)      2022-01-10   
4   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)      2022-01-10   

  PaiementDueDate  StaffID  Balance  SubTotalVAT1  Barcode  
0      2022-01-11        3     29.0          29.0  3000019  
1      2022-01-11        3     43.0          43.0  3000028  
2      2022-01-10        3     61.0          61.0  3000037  
3      2022-01-10        3     57.0          57.0  3000046  
4      2022-01-10        3     43.0          43.0  3000055

Notemos que Balance y SubTotalIVAT1 son iguales y estas a su vez también son iguales a Total, por lo que las eliminaremos

In [32]:
np.all(baInvoices_cleaned['SubTotalVAT1'] == baInvoices_cleaned['Balance'])

True

In [33]:
np.all(baInvoices_cleaned['Total'] == baInvoices_cleaned['Balance'])

True

In [34]:
baInvoices_cleaned.drop(['SubTotalVAT1', 'Balance'], axis=1, inplace=True)

In [35]:
baInvoices_cleaned.head()

InvoiceID  InvoiceNumber       Date Creation Date Invoice  Sub Total   
0          1              1 2022-01-11 10:15:06   2022-01-11       29.0  \
1          2              2 2022-01-11 10:29:02   2022-01-11       43.0   
2          3              3 2022-01-10 10:43:53   2022-01-10       61.0   
3          4              4 2022-01-10 10:44:34   2022-01-10       57.0   
4          5              5 2022-01-10 10:44:57   2022-01-10       43.0   

   Total                           Total letter DateDocDelivery   
0   29.0        (VEINTINUEVE PESOS 00/100 M.N.)      2022-01-11  \
1   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)      2022-01-11   
2   61.0       (SESENTA Y UN PESOS 00/100 M.N.)      2022-01-10   
3   57.0  (CINCUENTA Y SIETE PESOS 00/100 M.N.)      2022-01-10   
4   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)      2022-01-10   

  PaiementDueDate  StaffID  Barcode  
0      2022-01-11        3  3000019  
1      2022-01-11        3  3000028  
2      2022-01-10        3  3000037  
3      2022-01-10        3  3000046  
4      2022-01-10        3  3000055

Ahora notemos que **DateDocDelivery** y **PaimentDueDate** parecen repetir los mismos valores, veamos:

In [36]:
np.all(baInvoices_cleaned['DateDocDelivery'] == baInvoices_cleaned['PaiementDueDate'])

True

In [37]:
np.all(baInvoices_cleaned['Date Creation'].dt.date == baInvoices_cleaned['PaiementDueDate'])

True

Dado que ambas columnas se relacionan con **Date Creation**, entonces no serán relevantes y las dropeamos

In [38]:
baInvoices_cleaned.drop(['DateDocDelivery', 'PaiementDueDate'], axis=1, inplace=True)

In [39]:
baInvoices_cleaned

InvoiceID  InvoiceNumber       Date Creation Date Invoice  Sub Total   
0              1              1 2022-01-11 10:15:06   2022-01-11       29.0  \
1              2              2 2022-01-11 10:29:02   2022-01-11       43.0   
2              3              3 2022-01-10 10:43:53   2022-01-10       61.0   
3              4              4 2022-01-10 10:44:34   2022-01-10       57.0   
4              5              5 2022-01-10 10:44:57   2022-01-10       43.0   
...          ...            ...                 ...          ...        ...   
28749      28750          28740 2022-12-19 18:52:34   2022-12-19       43.0   
28750      28751          28741 2022-12-19 18:57:23   2022-12-19       69.0   
28751      28752          28742 2022-12-19 18:56:35   2022-12-19       67.0   
28752      28753          28743 2022-12-19 18:59:11   2022-12-19       65.0   
28753      28754          28744 2022-12-19 19:05:31   2022-12-19       50.0   

       Total                           Total letter  StaffID  Barcode  
0       29.0        (VEINTINUEVE PESOS 00/100 M.N.)        3  3000019  
1       43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        3  3000028  
2       61.0       (SESENTA Y UN PESOS 00/100 M.N.)        3  3000037  
3       57.0  (CINCUENTA Y SIETE PESOS 00/100 M.N.)        3  3000046  
4       43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        3  3000055  
...      ...                                    ...      ...      ...  
28749   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        2  3287505  
28750   69.0    (SESENTA Y NUEVE PESOS 00/100 M.N.)        3  3287514  
28751   67.0    (SESENTA Y SIETE PESOS 00/100 M.N.)        2  3287523  
28752   65.0    (SESENTA Y CINCO PESOS 00/100 M.N.)        2  3287532  
28753   50.0          (CINCUENTA PESOS 00/100 M.N.)        6  3287541  

[28754 rows x 9 columns]

#### Correctitud y Detección de Patrones

In [40]:
baInvoices_cleaned[baInvoices_cleaned['InvoiceID'] != baInvoices_cleaned['InvoiceNumber']]

InvoiceID  InvoiceNumber       Date Creation Date Invoice  Sub Total   
8214        8215           8214 2022-04-22 13:55:11   2022-04-22       46.0  \
8215        8216           8215 2022-04-22 14:02:32   2022-04-22       53.0   
8216        8217           8216 2022-04-22 14:07:38   2022-04-22      317.0   
8217        8218           8217 2022-04-22 14:13:26   2022-04-22       41.0   
8218        8219           8218 2022-04-22 14:30:42   2022-04-22       29.0   
...          ...            ...                 ...          ...        ...   
28749      28750          28740 2022-12-19 18:52:34   2022-12-19       43.0   
28750      28751          28741 2022-12-19 18:57:23   2022-12-19       69.0   
28751      28752          28742 2022-12-19 18:56:35   2022-12-19       67.0   
28752      28753          28743 2022-12-19 18:59:11   2022-12-19       65.0   
28753      28754          28744 2022-12-19 19:05:31   2022-12-19       50.0   

       Total                                Total letter  StaffID  Barcode  
8214    46.0         (CUARENTA Y SEIS PESOS 00/100 M.N.)        6  3082152  
8215    53.0        (CINCUENTA Y TRES PESOS 00/100 M.N.)        6  3082161  
8216   317.0  (TRESCIENTOS DIECISIETE PESOS 00/100 M.N.)        2  3082170  
8217    41.0           (CUARENTA Y UN PESOS 00/100 M.N.)        6  3082189  
8218    29.0             (VEINTINUEVE PESOS 00/100 M.N.)        2  3082198  
...      ...                                         ...      ...      ...  
28749   43.0         (CUARENTA Y TRES PESOS 00/100 M.N.)        2  3287505  
28750   69.0         (SESENTA Y NUEVE PESOS 00/100 M.N.)        3  3287514  
28751   67.0         (SESENTA Y SIETE PESOS 00/100 M.N.)        2  3287523  
28752   65.0         (SESENTA Y CINCO PESOS 00/100 M.N.)        2  3287532  
28753   50.0               (CINCUENTA PESOS 00/100 M.N.)        6  3287541  

[20540 rows x 9 columns]

Ahora lo que verificaremos será si la columna **Total letter** coincide con el número de **Total**

In [41]:
from num2words import num2words
import re
from unidecode import unidecode

In [42]:
precios = baInvoices_cleaned.loc[:,['Total', 'Total letter']]

In [43]:
validate = []
for i in precios.index.values:
    price_num = precios.loc[i][0]
    price_text = precios.loc[i][1]
    try:
        price_text = re.search(r'^(.*?)\bPESOS\b', price_text).group(1)
        price_text = price_text.strip()
    except:
        print(i)
    price_text = re.sub(r'[^A-Z\s]', '', price_text)
    price_num = num2words(price_num, lang='es', to='currency', separator=' ')
    price_num = re.search(r'^(.*?)\b euros\b', price_num).group(1).upper()
    price_num = unidecode(price_num)
    
    if price_num != price_text:
        print(price_num, price_text)
    else:
        validate.append(True)

In [44]:
np.all(validate)

True

Tenemos que tanto la parte numérica como la parte de texto del precio coincide.

In [45]:
baInvoices_cleaned

InvoiceID  InvoiceNumber       Date Creation Date Invoice  Sub Total   
0              1              1 2022-01-11 10:15:06   2022-01-11       29.0  \
1              2              2 2022-01-11 10:29:02   2022-01-11       43.0   
2              3              3 2022-01-10 10:43:53   2022-01-10       61.0   
3              4              4 2022-01-10 10:44:34   2022-01-10       57.0   
4              5              5 2022-01-10 10:44:57   2022-01-10       43.0   
...          ...            ...                 ...          ...        ...   
28749      28750          28740 2022-12-19 18:52:34   2022-12-19       43.0   
28750      28751          28741 2022-12-19 18:57:23   2022-12-19       69.0   
28751      28752          28742 2022-12-19 18:56:35   2022-12-19       67.0   
28752      28753          28743 2022-12-19 18:59:11   2022-12-19       65.0   
28753      28754          28744 2022-12-19 19:05:31   2022-12-19       50.0   

       Total                           Total letter  StaffID  Barcode  
0       29.0        (VEINTINUEVE PESOS 00/100 M.N.)        3  3000019  
1       43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        3  3000028  
2       61.0       (SESENTA Y UN PESOS 00/100 M.N.)        3  3000037  
3       57.0  (CINCUENTA Y SIETE PESOS 00/100 M.N.)        3  3000046  
4       43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        3  3000055  
...      ...                                    ...      ...      ...  
28749   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        2  3287505  
28750   69.0    (SESENTA Y NUEVE PESOS 00/100 M.N.)        3  3287514  
28751   67.0    (SESENTA Y SIETE PESOS 00/100 M.N.)        2  3287523  
28752   65.0    (SESENTA Y CINCO PESOS 00/100 M.N.)        2  3287532  
28753   50.0          (CINCUENTA PESOS 00/100 M.N.)        6  3287541  

[28754 rows x 9 columns]

### baInvoicesItems

#### Completitud

In [46]:
# Reporte completo
baInvoicesItems_completeness = completeness_report(baInvoicesItems)
print(baInvoicesItems_completeness.shape)
print(baInvoicesItems_completeness.index.values)

(35, 2)
['InvoiceItemID' 'InvoiceID' 'Quantity' 'StockID' 'Description'
 'Price Unit' 'Total' 'CostPrice' 'DescriptionJustifyOrigin' 'Language'
 'Key' 'LineNbr' 'VATPerc' 'VersionSync' 'BranchID' 'Branch_InvoiceItemID'
 'Master_InvoiceItemID' 'ProjectIDOLD' 'SerialNumber' 'MustBeDeleted'
 'SerialFrom' 'SerialTo' 'NumeroDePedimento' 'FechaDePedimento'
 'NombreDeAduana' 'VATRetentionPerc' 'IEPSPerc' 'Comments' 'DiscountPerc'
 'Unit' 'CommentsAdic' 'DocumentItemIDServer' 'IVASobreIEPS' 'IVAExento'
 'ProjectID']


In [47]:
baInvoicesItems_completeness_1 = baInvoicesItems_completeness[baInvoicesItems_completeness['Completeness'] == 1]
print(baInvoicesItems_completeness_1.shape)
print(baInvoicesItems_completeness_1.index.values)

(22, 2)
['InvoiceItemID' 'InvoiceID' 'Quantity' 'StockID' 'Description'
 'CostPrice' 'Language' 'LineNbr' 'VATPerc' 'VersionSync' 'BranchID'
 'Branch_InvoiceItemID' 'Master_InvoiceItemID' 'ProjectIDOLD'
 'MustBeDeleted' 'VATRetentionPerc' 'IEPSPerc' 'DiscountPerc'
 'DocumentItemIDServer' 'IVASobreIEPS' 'IVAExento' 'ProjectID']


In [48]:
baInvoicesItems_completeness_0 = baInvoicesItems_completeness[baInvoicesItems_completeness['Completeness'] == 0]
print(baInvoicesItems_completeness_0.shape)
print(baInvoicesItems_completeness_0.index.values)

(11, 2)
['DescriptionJustifyOrigin' 'Key' 'SerialNumber' 'SerialFrom' 'SerialTo'
 'NumeroDePedimento' 'FechaDePedimento' 'NombreDeAduana' 'Comments' 'Unit'
 'CommentsAdic']


In [49]:
baInvoicesItems_miss = baInvoicesItems_completeness[(baInvoicesItems_completeness['Missing'] != 0) & (baInvoicesItems_completeness['Missing'] != 1)]
print(baInvoicesItems_miss.shape)
print(baInvoicesItems_miss.index.values)

(2, 2)
['Price Unit' 'Total']


In [50]:
baInvoicesItems[baInvoicesItems['Price Unit'].isnull() == True]

InvoiceItemID  InvoiceID  Quantity  StockID   
37787          37788      21371         1     1983  \

                                   Description  Price Unit  Total  CostPrice   
37787  fruti mich fresa/chamoy salsa mix 1,1kg         NaN    NaN        0.0  \

       DescriptionJustifyOrigin  Language  ...  VATRetentionPerc  IEPSPerc   
37787                       NaN         3  ...                 0         0  \

       Comments  DiscountPerc  Unit  CommentsAdic  DocumentItemIDServer   
37787       NaN             0   NaN           NaN                     0  \

       IVASobreIEPS  IVAExento  ProjectID  
37787         False      False          0  

[1 rows x 35 columns]

In [51]:
# Eliminamos los registros que no tienen precio unitario
baInvoicesItems_cleaned = baInvoicesItems.drop(37787, axis=0)
# Eliminamos a su vez las columnas que no aportan información y que su completitud es 0
baInvoicesItems_cleaned = baInvoicesItems_cleaned.drop(baInvoicesItems_completeness_0.index.values, axis=1)

Las siguientes son las columnas más relevantes que tenemos en esta tabla

In [52]:
baInvoicesItems_cleaned = drop_equal_values(baInvoicesItems_cleaned, 0.8)
baInvoicesItems_cleaned

InvoiceItemID  InvoiceID  StockID   
0                  1          1        5  \
1                  2          2       78   
2                  3          2       66   
3                  4          3     2568   
4                  5          4      732   
...              ...        ...      ...   
50309          50310      28750      228   
50310          50311      28751       62   
50311          50312      28752     2490   
50312          50313      28753      826   
50313          50314      28754     2642   

                                   Description  Price Unit  Total  CostPrice   
0               de la rosa bianchi grande 45pz      29.000   29.0   22.50000  \
1                 reyma 7oz vaso plastico 50pz      30.000   30.0   23.40000   
2                classy cuchara pastelera 50pz      13.000   13.0    8.83000   
3                             macma deju 280gr      61.000   61.0   48.60000   
4                          macma escoces 330gr      57.000   57.0   45.49000   
...                                        ...         ...    ...        ...   
50309             12x25 celofan adhesivo 100pz      43.000   43.0   27.00000   
50310          25x35 pb hoja mediana polipapel       0.069   69.0    0.05458   
50311                ricolino kranky mini 25pz      67.000   67.0   52.90000   
50312            macma selecciones chico 330gr      65.000   65.0    0.00000   
50313  meltblown cubrebocas tricapa 50pz negro      50.000   50.0   22.50000   

       LineNbr  
0            1  
1            1  
2            2  
3            1  
4            1  
...        ...  
50309        1  
50310        1  
50311        1  
50312        1  
50313        1  

[50313 rows x 8 columns]

### baStaff

#### Completitud

In [53]:
# Reporte completo
baStaff_completeness = completeness_report(baStaff)
print(baStaff_completeness.shape)
print(baStaff_completeness.index.values)

(64, 2)
['StaffID' 'Name' 'Hours' 'Function' 'Title' 'Address' 'Tel' 'Fax'
 'Mobile' 'EMail' 'Commission' 'Salary' 'SalaryCurrency' 'SalaryPeriod'
 'Comments' 'VATNo' 'QuotationTel' 'CommissionOn' 'HoursMonday'
 'HoursTuesday' 'HoursWednesday' 'HoursThursday' 'HoursFriday'
 'HoursSaturday' 'HoursSunday' 'BranchID' 'Branch_StaffID'
 'MustBeSynchronized' 'Deleted' 'UserID' 'VersionSync' 'CardID'
 'PictureSourceDocument' 'Department' 'StaffNo' 'ProfileID' 'ClientID'
 'ProjectID' 'ScanTelcelPhoneNumber' 'FechaBaja' 'CURP' 'IMSS' 'FechaAlta'
 'StatusActivity' 'TipoRegimen' 'TipoContrato' 'TipoJornada' 'TipoRiesgo'
 'SalarioDiarioIntegrado' 'SalarioBaseCotAport' 'Banco' 'CLABE'
 'AddressStreet1' 'AddressNoExt' 'AddressNoInt' 'AddressStreet2'
 'AddressCity' 'AddressZipCode' 'AddressState' 'AddressBetweenStreets'
 'AddressLocality' 'AddressCountry' 'AddressGLN' 'MetodoDePago']


In [54]:
baStaff_completeness_1 = baStaff_completeness[baStaff_completeness['Completeness'] == 1]
print(baStaff_completeness_1.shape)
print(baStaff_completeness_1.index.values)

(16, 2)
['StaffID' 'Name' 'Hours' 'Function' 'CommissionOn' 'Branch_StaffID'
 'MustBeSynchronized' 'Deleted' 'VersionSync' 'ProfileID' 'StatusActivity'
 'TipoRegimen' 'TipoRiesgo' 'SalarioDiarioIntegrado' 'SalarioBaseCotAport'
 'Banco']


In [55]:
baStaff_completeness_0 = baStaff_completeness[baStaff_completeness['Completeness'] == 0]
print(baStaff_completeness_0.shape)
print(baStaff_completeness_0.index.values)

(47, 2)
['Title' 'Address' 'Tel' 'Fax' 'Mobile' 'EMail' 'Commission' 'Salary'
 'SalaryCurrency' 'SalaryPeriod' 'VATNo' 'QuotationTel' 'HoursMonday'
 'HoursTuesday' 'HoursWednesday' 'HoursThursday' 'HoursFriday'
 'HoursSaturday' 'HoursSunday' 'BranchID' 'UserID' 'CardID'
 'PictureSourceDocument' 'Department' 'StaffNo' 'ClientID' 'ProjectID'
 'ScanTelcelPhoneNumber' 'FechaBaja' 'CURP' 'IMSS' 'FechaAlta'
 'TipoContrato' 'TipoJornada' 'CLABE' 'AddressStreet1' 'AddressNoExt'
 'AddressNoInt' 'AddressStreet2' 'AddressCity' 'AddressZipCode'
 'AddressState' 'AddressBetweenStreets' 'AddressLocality' 'AddressCountry'
 'AddressGLN' 'MetodoDePago']


In [56]:
baStaff_miss = baStaff_completeness[(baStaff_completeness['Missing'] != 0) & (baStaff_completeness['Missing'] != 1)]
print(baStaff_miss.shape)
print(baStaff_miss.index.values)

(1, 2)
['Comments']


In [57]:
baStaff[['Title','Address','Tel','Fax','Mobile','EMail','Commission','Salary'
,'SalaryCurrency','SalaryPeriod','VATNo','QuotationTel','HoursMonday'
,'HoursTuesday','HoursWednesday']]

Title  Address  Tel  Fax  Mobile  EMail  Commission  Salary   
0    NaN      NaN  NaN  NaN     NaN    NaN         NaN     NaN  \
1    NaN      NaN  NaN  NaN     NaN    NaN         NaN     NaN   
2    NaN      NaN  NaN  NaN     NaN    NaN         NaN     NaN   
3    NaN      NaN  NaN  NaN     NaN    NaN         NaN     NaN   

   SalaryCurrency  SalaryPeriod  VATNo  QuotationTel  HoursMonday   
0             NaN           NaN    NaN           NaN          NaN  \
1             NaN           NaN    NaN           NaN          NaN   
2             NaN           NaN    NaN           NaN          NaN   
3             NaN           NaN    NaN           NaN          NaN   

   HoursTuesday  HoursWednesday  
0           NaN             NaN  
1           NaN             NaN  
2           NaN             NaN  
3           NaN             NaN

In [58]:
baStaff['Comments']

0                    inventario inicial 27/07/20
1                                            NaN
2    sandibel marin hernandez_x000d_\n21 octubre
3                                            NaN
Name: Comments, dtype: object

In [59]:
baStaff[['StaffID','Name','Hours','Function','CommissionOn','Branch_StaffID'
,'MustBeSynchronized','Deleted','VersionSync','ProfileID','StatusActivity'
,'TipoRegimen','TipoRiesgo','SalarioDiarioIntegrado','SalarioBaseCotAport'
,'Banco']]

StaffID      Name  Hours  Function  CommissionOn  Branch_StaffID   
0        1       UNO      0         4             0               0  \
1        2  MARYCRUZ      0         4             0               0   
2        3    CARMEN      0         4             0               0   
3        6    LORENA      0         4             0               0   

   MustBeSynchronized  Deleted  VersionSync  ProfileID  StatusActivity   
0                True    False           76          0               0  \
1                True    False          125          0               0   
2                True    False          121          0               0   
3                True    False          120          0               0   

   TipoRegimen  TipoRiesgo  SalarioDiarioIntegrado  SalarioBaseCotAport  Banco  
0            0           0                       0                    0      0  
1            0           0                       0                    0      0  
2            0           0                       0                    0      0  
3            0           0                       0                    0      0

In [60]:
baStaffClean = baStaff.drop(baStaff_completeness_0.index.values,axis=1)
baStaffClean = baStaffClean.drop(baStaffClean.columns[baStaffClean.eq(0).all()].values,axis=1)

In [61]:
baStaffClean

StaffID      Name  Function                                     Comments   
0        1       UNO         4                  inventario inicial 27/07/20  \
1        2  MARYCRUZ         4                                          NaN   
2        3    CARMEN         4  sandibel marin hernandez_x000d_\n21 octubre   
3        6    LORENA         4                                          NaN   

   MustBeSynchronized  VersionSync  
0                True           76  
1                True          125  
2                True          121  
3                True          120

### baStock

#### Completitud

In [62]:
# Reporte completo
baStock_completeness = completeness_report(baStock)
print(baStock_completeness.shape)
print(baStock_completeness.index.values)

(106, 2)
['StockID' 'StockIDServer' 'Key' 'Familly' 'Group' 'Item' 'Description'
 'CostPrice' 'QtyMinimum' 'DateValidity' 'Waranty' 'EndUserItem' 'Deleted'
 'JustifyOrigin' 'Uploaded' 'EBusinessUpload' 'EBusinessNewProduct'
 'EBusinessPromotion' 'EBusinessPrice' 'EbusinessPromotionPrice' 'BarCode'
 'PriceList' 'Service' 'PictureSourceDocument' 'MustBeSynchronized'
 'Package' 'PercCommission' 'Discontinued' 'SortNo' 'PrintBarCode'
 'Volume' 'CustomRate' 'DTA' 'UserID' 'SoldInStore' 'QtyPresWareHouse'
 'QtyPresStores' 'QtyPresFranchises' 'QtyPresQuarantine' 'QtyPresTotal'
 'QtyTransit' 'QtyPendSuppliers' 'QtyPendProduction' 'QtyPendTotal'
 'PendingFrom' 'PendingTo' 'PendingFinalDate' 'QtyTOTAL' 'QtyToDeliver'
 'QtyAvailable' 'QtyAvailableForStore' 'DateAvailableForStore'
 'UpdateStockQty' 'Level' 'Unit' 'Rotation' 'CurrencyPriceList'
 'SalesAveragePerWeek' 'SalesNbrWeeksInfo' 'Matrix' 'MatrixCriteria1'
 'MatrixCriteria2' 'MatrixStockIDParent' 'VATPerc' 'UploadImageToInternet'
 'WeeksToDe

In [63]:
baStock_completeness_1 = baStock_completeness[baStock_completeness['Completeness'] == 1]
print(baStock_completeness_1.shape)
print(baStock_completeness_1.index.values)

(73, 2)
['StockID' 'Item' 'CostPrice' 'QtyMinimum' 'EndUserItem' 'Deleted'
 'JustifyOrigin' 'Uploaded' 'EBusinessUpload' 'EBusinessNewProduct'
 'EBusinessPromotion' 'EBusinessPrice' 'EbusinessPromotionPrice' 'BarCode'
 'PriceList' 'Service' 'MustBeSynchronized' 'Package' 'PercCommission'
 'Discontinued' 'SortNo' 'PrintBarCode' 'CustomRate' 'DTA' 'SoldInStore'
 'QtyPresWareHouse' 'QtyPresStores' 'QtyPresFranchises'
 'QtyPresQuarantine' 'QtyPresTotal' 'QtyTransit' 'QtyPendSuppliers'
 'QtyPendProduction' 'QtyPendTotal' 'PendingFinalDate' 'QtyTOTAL'
 'QtyToDeliver' 'QtyAvailable' 'QtyAvailableForStore' 'UpdateStockQty'
 'Level' 'CurrencyPriceList' 'SalesAveragePerWeek' 'SalesNbrWeeksInfo'
 'Matrix' 'MatrixCriteria1' 'MatrixCriteria2' 'VATPerc'
 'UploadImageToInternet' 'WeeksToDeliverSupplier' 'LastMonthSales'
 'CostPriceAverage' 'VersionSync' 'AverageSaleQty' 'CanSaleNegative'
 'PriceListSupplier' 'Retention' 'PromoType' 'QtyToDeliverPosterior'
 'SalePriceType' 'SupplierDiscount' 'Supplier

In [64]:
baStock_completeness_0 = baStock_completeness[baStock_completeness['Completeness'] == 0]
print(baStock_completeness_0.shape)
print(baStock_completeness_0.index.values)

(19, 2)
['StockIDServer' 'Key' 'Waranty' 'PictureSourceDocument' 'Volume'
 'PendingFrom' 'PendingTo' 'DateAvailableForStore' 'Rotation'
 'MatrixStockIDParent' 'Category1' 'Category2' 'CountryOriginID'
 'SupplierID' 'BuyingUnits' 'CPExecuteFunction' 'TelProductCategoryID'
 'WSQuantities' 'URL']


In [65]:
baStock_miss = baStock_completeness[(baStock_completeness['Missing'] != 0) & (baStock_completeness['Missing'] != 1)]
print(baStock_miss.shape)
print(baStock_miss.index.values)

(14, 2)
['Familly' 'Group' 'Description' 'DateValidity' 'UserID' 'Unit'
 'AverageSaleNbrWeeks' 'WeeksOnMarket' 'LastSale' 'QtyMaximum'
 'TelCategoryActivatedID' 'CPAssignSaleOtherBranchID' 'eWalletPerc'
 'WSServerProductID']


In [66]:
baStock[baStock_miss.index.values]

Familly Group Description DateValidity  UserID  Unit   
0        DE LA ROSA   NaN         NaN          NaT     NaN   NaN  \
1        DE LA ROSA   NaN         NaN          NaT     NaN   NaN   
2        DE LA ROSA   NaN         NaN          NaT     NaN   NaN   
3        DE LA ROSA   NaN         NaN          NaT     NaN   NaN   
4        DE LA ROSA   NaN         NaN          NaT     NaN   NaN   
...             ...   ...         ...          ...     ...   ...   
3036  CASA FIGUEROA   NaN         NaN          NaT     NaN   NaN   
3037         PAYASO   NaN         NaN          NaT     NaN   NaN   
3038     DE LA ROSA   NaN         NaN          NaT     NaN   NaN   
3039        CUETARA   NaN         NaN          NaT     NaN   NaN   
3040         ALTESA   NaN         NaN          NaT     NaN   NaN   

      AverageSaleNbrWeeks  WeeksOnMarket  LastSale  QtyMaximum   
0                    28.0            NaN       0.0         NaN  \
1                     NaN            NaN       NaN         NaN   
2                    28.0            NaN       1.0         NaN   
3                     NaN            NaN       NaN         NaN   
4                    28.0            NaN       0.0         NaN   
...                   ...            ...       ...         ...   
3036                  0.0            0.0       0.0         0.0   
3037                  0.0            0.0       0.0         0.0   
3038                  0.0            0.0       0.0         0.0   
3039                  0.0            0.0       0.0         0.0   
3040                  0.0            0.0       0.0         0.0   

      TelCategoryActivatedID  CPAssignSaleOtherBranchID  eWalletPerc   
0                        NaN                        NaN          0.0  \
1                        NaN                        NaN          0.0   
2                        NaN                        NaN          0.0   
3                        NaN                        NaN          0.0   
4                        NaN                        NaN          0.0   
...                      ...                        ...          ...   
3036                     0.0                        0.0          0.0   
3037                     0.0                        0.0          0.0   
3038                     0.0                        0.0          0.0   
3039                     0.0                        0.0          0.0   
3040                     0.0                        0.0          0.0   

      WSServerProductID  
0                   NaN  
1                   NaN  
2                   NaN  
3                   NaN  
4                   NaN  
...                 ...  
3036                0.0  
3037                0.0  
3038                0.0  
3039                0.0  
3040                0.0  

[3041 rows x 14 columns]

In [67]:
baStock['WSServerProductID'].unique()

array([nan,  0.])

In [68]:
baStock[baStock_completeness_1.index.values].head(2)

StockID                                          Item  CostPrice   
0        1  de la rosa bianchi choco amarillo 50pz 350gr       47.4  \
1        2   CAJA de la rosa bianchi choco amarillo c/20      948.0   

   QtyMinimum  EndUserItem  Deleted  JustifyOrigin  Uploaded  EBusinessUpload   
0           1         True    False          False     False            False  \
1           0         True    False          False     False            False   

   EBusinessNewProduct  ...  Compra  Requisicion  Inventariable  Terminado   
0                False  ...   False        False          False      False  \
1                False  ...   False        False          False      False   

   InsumoBase   CommentWhenAvailableForStore  CanOnlyInvoiceProduct   
0       False      2 disponible de inmediato                  False  \
1       False  NO DISPONIBLE (NO PROGRAMADO)                  False   

   CanOnlyTicketProduct  IVASobreIEPS  IVAExento  
0                 False         False      False  
1                 False         False      False  

[2 rows x 73 columns]

#### Limpieza

In [69]:
baStock_cleaned = baStock.drop(baStock_completeness_0.index.values,axis=1)

In [70]:
baStock_cleaned.shape

(3041, 87)

In [71]:
cols_to_drop = [col for col in baStock_cleaned.columns if ((baStock_cleaned[col].notnull().any()) and np.array_equal(baStock_cleaned[col].dropna().unique(), np.array([0.])) and np.any(baStock_cleaned[col].isnull()))]
baStock_cleaned = baStock_cleaned.drop(cols_to_drop, axis = 1)

In [72]:
baStock_cleaned.shape

(3041, 81)

In [73]:
baStock_cleaned = drop_equal_values(baStock_cleaned, .8)

In [74]:
baStock_cleaned.shape

(3041, 24)

In [75]:
# Calcular el porcentaje de valores no nulos en cada columna
porcentaje_no_nulos = baStock_cleaned.notnull().sum() / len(baStock_cleaned)

# Identificar las columnas que tienen un porcentaje de valores no nulos menor al 80%
columnas_a_eliminar = porcentaje_no_nulos[porcentaje_no_nulos <= 0.8].index.tolist()

# Eliminar las columnas del DataFrame
baStock_cleaned = baStock_cleaned.drop(columnas_a_eliminar, axis=1)

In [76]:
baStock_cleaned.shape

(3041, 17)

In [77]:
baStock_cleaned.head(2)

StockID     Familly                                          Item   
0        1  DE LA ROSA  de la rosa bianchi choco amarillo 50pz 350gr  \
1        2  DE LA ROSA   CAJA de la rosa bianchi choco amarillo c/20   

   CostPrice  QtyMinimum  BarCode  PriceList  Package  QtyPresStores   
0       47.4           1       17       60.0    False              2  \
1      948.0           0       24     1185.0     True              0   

   QtyPresTotal  QtyTOTAL  QtyAvailable  QtyAvailableForStore  Level   
0             2         2             2                     7      2  \
1             0         0             0                     0      1   

   VersionSync  AverageSaleQty   CommentWhenAvailableForStore  
0       168663           876.0      2 disponible de inmediato  
1       171420             0.0  NO DISPONIBLE (NO PROGRAMADO)

In [78]:
for col in baStock_cleaned.columns:
    print(col, len(baStock_cleaned[col].unique()))

StockID 3041
Familly 326
Item 3041
CostPrice 1483
QtyMinimum 22
BarCode 3041
PriceList 759
Package 2
QtyPresStores 73
QtyPresTotal 73
QtyTOTAL 73
QtyAvailable 73
QtyAvailableForStore 73
Level 2
VersionSync 3041
AverageSaleQty 327
CommentWhenAvailableForStore 88


Todos aquellos que son BULTOS o CAJAS podemos decir que vienen en paquetes.

In [79]:
# x = baStock_cleaned[baStock_cleaned['Package']== True]
# # check if item contains "CAJA"
# result = x[x['Item'].str.contains("CAJA", na=False)]
# idx = result.index
# x[~x.index.isin(idx)]

In [80]:
baInvoices_cleaned

InvoiceID  InvoiceNumber       Date Creation Date Invoice  Sub Total   
0              1              1 2022-01-11 10:15:06   2022-01-11       29.0  \
1              2              2 2022-01-11 10:29:02   2022-01-11       43.0   
2              3              3 2022-01-10 10:43:53   2022-01-10       61.0   
3              4              4 2022-01-10 10:44:34   2022-01-10       57.0   
4              5              5 2022-01-10 10:44:57   2022-01-10       43.0   
...          ...            ...                 ...          ...        ...   
28749      28750          28740 2022-12-19 18:52:34   2022-12-19       43.0   
28750      28751          28741 2022-12-19 18:57:23   2022-12-19       69.0   
28751      28752          28742 2022-12-19 18:56:35   2022-12-19       67.0   
28752      28753          28743 2022-12-19 18:59:11   2022-12-19       65.0   
28753      28754          28744 2022-12-19 19:05:31   2022-12-19       50.0   

       Total                           Total letter  StaffID  Barcode  
0       29.0        (VEINTINUEVE PESOS 00/100 M.N.)        3  3000019  
1       43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        3  3000028  
2       61.0       (SESENTA Y UN PESOS 00/100 M.N.)        3  3000037  
3       57.0  (CINCUENTA Y SIETE PESOS 00/100 M.N.)        3  3000046  
4       43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        3  3000055  
...      ...                                    ...      ...      ...  
28749   43.0    (CUARENTA Y TRES PESOS 00/100 M.N.)        2  3287505  
28750   69.0    (SESENTA Y NUEVE PESOS 00/100 M.N.)        3  3287514  
28751   67.0    (SESENTA Y SIETE PESOS 00/100 M.N.)        2  3287523  
28752   65.0    (SESENTA Y CINCO PESOS 00/100 M.N.)        2  3287532  
28753   50.0          (CINCUENTA PESOS 00/100 M.N.)        6  3287541  

[28754 rows x 9 columns]

In [81]:
baInvoicesItems_cleaned

InvoiceItemID  InvoiceID  StockID   
0                  1          1        5  \
1                  2          2       78   
2                  3          2       66   
3                  4          3     2568   
4                  5          4      732   
...              ...        ...      ...   
50309          50310      28750      228   
50310          50311      28751       62   
50311          50312      28752     2490   
50312          50313      28753      826   
50313          50314      28754     2642   

                                   Description  Price Unit  Total  CostPrice   
0               de la rosa bianchi grande 45pz      29.000   29.0   22.50000  \
1                 reyma 7oz vaso plastico 50pz      30.000   30.0   23.40000   
2                classy cuchara pastelera 50pz      13.000   13.0    8.83000   
3                             macma deju 280gr      61.000   61.0   48.60000   
4                          macma escoces 330gr      57.000   57.0   45.49000   
...                                        ...         ...    ...        ...   
50309             12x25 celofan adhesivo 100pz      43.000   43.0   27.00000   
50310          25x35 pb hoja mediana polipapel       0.069   69.0    0.05458   
50311                ricolino kranky mini 25pz      67.000   67.0   52.90000   
50312            macma selecciones chico 330gr      65.000   65.0    0.00000   
50313  meltblown cubrebocas tricapa 50pz negro      50.000   50.0   22.50000   

       LineNbr  
0            1  
1            1  
2            2  
3            1  
4            1  
...        ...  
50309        1  
50310        1  
50311        1  
50312        1  
50313        1  

[50313 rows x 8 columns]

In [82]:
baInvoices_cleaned.to_csv('baInvoices_cleaned.csv', index=False)
baInvoicesItems_cleaned.to_csv('baInvoicesItems_cleaned.csv', index=False)
baStock_cleaned.to_csv('baStock_cleaned.csv', index=False)
baStaffClean.to_csv('baStaff_cleaned.csv', index=False)